File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
import math
def isPrime(target):
    if target<=1:
        return False
    if target==2:
        return True
    for i in range(2, int(math.sqrt(target))+1):
        if target%i==0:
            return False
    return True
# def DFS(L, max_level):
#     if L==max_level:
#         return
    
#     DFS(L+1)
    
# def solution(numbers):
#     ## 완탐 가능할듯
#     answer=set()
#     cand=[num for num in numbers]
#     DFS(0, len(cand))
#     ## 순열로 소수인지 파악
#     return len(answer)
def solution(numbers):
    ## 완탐 가능할듯
    answer=set()
    cand=[num for num in numbers]

    ## 순열로 소수인지 파악
    for i in range(1, len(cand)+1):
        for j in permutations(cand, i):
            tmp=int(''.join(j))
            if isPrime(tmp):
                print(tmp)
                answer.add(tmp)
    return len(answer)

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
class Solution {
    static Set<Integer> res=new HashSet<>();
    static boolean[] visited;
    public int solution(String numbers) {
        // ## DFS로 만들 수 있는 수를 모두 구하고 소수구하기
        int n=numbers.length();
        visited=new boolean[n];
        DFS(numbers, "", n);
        return isPrime();

    }
    public void DFS(String numbers, String number, int n){
        if (number!=""){
            res.add(Integer.valueOf(number));
        }
        for (int i=0; i<n; i++){
            if (visited[i]==false){
                visited[i]=true;

                DFS(numbers, number+numbers.charAt(i), n);
                visited[i]=false;
            }
        }    
    }
    public int isPrime(){
        int cnt=0;

        for (int target : res){
            if (target<2){
                continue;
            }
            boolean flag=true;
            for (int i=2; i<(int)Math.sqrt(target)+1; i++){
                if (target%i==0){
                    flag=false;
                    break;
                }
            }
            if (flag==true){
                cnt+=1;
            }
        }
        return cnt;
    }
}
